SIP-14 improvements

viktorklang opened the pull request:
Add missing canonical combinators:
- `def transform[S](f: Try[T] => Try[S])(implicit executor: ExecutionContext): Future[S]`
- `def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S]`
- `def flatten[S](implicit ev: T <:< Future[S]): Future[S]`
- `def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R]`

Add missing utilities:
-  `val unit: Future[Unit]` in `object Future`
-  `val never: Future[Nothing]` in `object Future`
-  `def defaultBlockContext: BlockContext` in `object BlockContext`

Refactors:
- the `scala.concurrent.Future` trait to not explicit create any `Promises`,
  so that implementations can control implementation type,
  this is mainly facilitated through adding of the `transform` and `transformWith` methods.
- the implementation of `ExecutionContextImpl` has been cleaned up

Securing:
- Add a self-check in `completeWith` and `tryCompleteWith` to avoid cycles in trait Promise
- Capping the maximum number of threads for the global `ExecutionContext` to the max parallelism
- Implementing (almost) all `Future` combinators on `transformWith` and `transform` means
   that `DefaultPromise` linking works on both `(flat)map` and `recover(With)`
- Nested `blocking {}` should not spawn extra threads beyond the first.

Removes:
- the private `internalExecutor` method in favor of an import in trait `Future`
- the private `internalExecutor` method in favor of an import in trait `Promise`
- the `AtomicReferenceFieldUpdater` in `AbstractPromise` since we're using `Unsafe`
- `scala.concurrent.impl.Future` is no longer needed

Deprecates:
- `Future.onSuccess` - discourage the use of callbacks
                       (and is also redundant considering `foreach` and `onComplete`)
- `Future.onFailure` - discourage the use of callbacks
                     (and is also redundant considering `onComplete` and `failed.foreach`)
- `ExecutionContext.prepare` - it was ill specced and it is too easy to forget to call it
                   (or even know when to call it or call it more times than needed)

Reimplements:
- `failed` - in terms of `transform`
- `map` - in terms of `transform`
- `flatMap` - in terms of `transformWith`
- `recover` - in terms of `transform`
- `recoverWith` - in terms of `transformWith`
- `zip` - in terms of `flatMap` + `map`
- `fallbackTo` - in terms of `recoverWith` + `recoverWith`
- `andThen` - in terms of `transform`

Miscellaneous:
- Giving the threads of `ExecutionContext.global` sensible names

Documentation:
- Almost all methods on `Future` and `Promise` have been revisited and had their ScalaDoc updated

Tests:
- Yes

viktorklang committed:
Add missing canonical combinators:
  - `def transform[S](f: Try[T] => Try[S])(implicit executor: ExecutionContext): Future[S]`
  - `def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S]`
  - `def flatten[S](implicit ev: T <:< Future[S]): Future[S]`
  - `def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R]`

Add missing utilities:
  -  `val unit: Future[Unit]` in `object Future`
  -  `object never extends Future[Nothing]` in `object Future`
  -  `def defaultBlockContext: BlockContext` in `object BlockContext`
  -  `def toString: String` on stdlib implementations of `Future`

Refactors:
  - the `scala.concurrent.Future` trait to not explicit create any `Promises`,
    so that implementations can control implementation type,
    this is mainly facilitated through adding of the `transform` and `transformWith` methods.
  - the implementation of `ExecutionContextImpl` has been cleaned up
  - the `scala.concurrent.impl.DefaultPromise` has been reimplemented to not use `sun.misc.Unsafe`

Securing:
  - Add a self-check in `completeWith` and `tryCompleteWith` to avoid cycles in trait Promise
  - Capping the maximum number of threads for the global `ExecutionContext` to the max parallelism
  - Implementing (almost) all `Future` combinators on `transformWith` and `transform` means
     that `DefaultPromise` linking works on both `(flat)map` and `recover(With)`
  - Nested `blocking {}` should not spawn extra threads beyond the first.

Removes:
  - the private `internalExecutor` method in favor of an import in trait `Future`
  - the private `internalExecutor` method in favor of an import in trait `Promise`
  - the `AtomicReferenceFieldUpdater` in `AbstractPromise` since we're using `Unsafe`
  - `scala.concurrent.impl.Future` is no longer needed

Deprecates:
  - `Future.onSuccess` - discourage the use of callbacks
                         (and is also redundant considering `foreach` and `onComplete`)
  - `Future.onFailure` - discourage the use of callbacks
                       (and is also redundant considering `onComplete` and `failed.foreach`)
  - `ExecutionContext.prepare` - it was ill specced and it is too easy to forget to call it
                     (or even know when to call it or call it more times than needed)
  - All classes in scala.concurrent.forkjoin. Scala 2.12 will be Java 8+ and as such the jsr166e
    should be used as included in java.util.concurrent.

Reimplements:
  - `failed` - in terms of `transform`
  - `map` - in terms of `transform`
  - `flatMap` - in terms of `transformWith`
  - `recover` - in terms of `transform`
  - `recoverWith` - in terms of `transformWith`
  - `zip` - in terms of `flatMap` + `map`
  - `fallbackTo` - in terms of `recoverWith` + `recoverWith`
  - `andThen` - in terms of `transform`

Miscellaneous:
  - Giving the threads of `ExecutionContext.global` sensible names
  - Optimizes `object Future.successful` and `object Future.failed` are now separate implementations,
    to optimize for the result, avoiding doing work for the "other branch".
  - Optimizes `compressedRoot()` by avoiding double-calls to volatile get.

Documentation:
  - Almost all methods on `Future` and `Promise` have been revisited and had their ScalaDoc updated

Tests:
  - Yes

retronym commented:
Historical context:

Call for feedback and review:
https://twitter.com/viktorklang/status/515536221276483585
https://groups.google.com/forum/#!topic/scala-sips/bZs6919Plo4

Review comments from a draft pull request:
https://github.com/viktorklang/scala/pull/7

retronym commented:
I'd like to hitch https://issues.scala-lang.org/browse/SI-8849 on to this change (weakening the type of `Implicits.global` so that it doesn't silently win the implicit specificity battle with an implicit method param of type `ExecutionContext`. This will require prominent release noting, as it breaks our source compatibility policy, but the status quo is pretty horrible.

retronym commented:
I've updated the PR description with prettier markdown. Could you please transfer that to the commit comment?

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -60,6 +57,10 @@ import scala.reflect.ClassTag
  *
+ *  @define swallowsExceptions
+ *  Since this method executes asynchronously and does not produce a return value,
+ *  any non-fatal exceptions thrown will be swallowed.
```
What is the interaction, if any, with the thread's uncaught exception handler?

viktorklang replied:
Technically they'll be propagated to the EC's reportFailure method, that's probably a better text, will fix: https://github.com/viktorklang/scala/blob/wip-sip14-updates-%E2%88%9A/src/library/scala/concurrent/Future.scala#L471

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -376,24 +408,24 @@ trait Future[+T] extends Awaitable[T] {
-      }
-    }
-    p.future
+    recoverWith { case _ => that } recoverWith { case _ => this }
```
Beautiful!

viktorklang replied:
Thank you!

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -534,6 +646,51 @@ object Future {
+   * @param op       the fold operation to be applied to the zero and futures
+   * @return         the `Future` holding the result of the fold
+   */
+  def foldLeft[T, R](futures: scala.collection.immutable.Iterable[Future[T]])(zero: R)(op: (R, T) => R)(implicit executor: ExecutionContext): Future[R] = {
```
Could you comment on the somewhat restrictive choice of `immutable.Iterable` here (as compared with `TraversableOnce` in `traverse`)?

viktorklang replied:
Yes, it is because it is traversed asynchronously.

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -534,6 +646,51 @@ object Future {
+   *
+   * @tparam T       the type of the value of the input Futures
+   * @tparam R       the type of the value of the returned `Future`
+   * @param futures  the `scala.collection.immutable.Iterable` of Futures to be folded
```
Let's avoid repeating the type signatures here, I'd prefer `"the futures to be folded"

viktorklang commented:
I'm OK with fixing https://issues.scala-lang.org/browse/SI-8849 in this PR as well

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -534,6 +646,51 @@ object Future {
+      if (!i.hasNext) successful(prevValue)
+      else i.next().flatMap { value => foldNext(i, op(prevValue, value)) }
+
+    foldNext(futures.iterator, zero)
```
Looks good, it will allow the futures that have been completed to collected as we fold (assuming the input `Iterable`'s iterator supports that.)

e.g.

```
scala> Stream.continually(Array.ofDim[Double](10000)).take(1000000).iterator.foldLeft(0)((x, y) => 0)
res9: Int = 0 
```

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -553,14 +718,42 @@ object Future {
+
+    val i = futures.iterator
+    if (!i.hasNext) failed(new NoSuchElementException("reduceLeft attempted on empty collection"))
+    else i.next() flatMap { v => reduceNext(i, v) }
```
We could share this code with `foldLeft`.

viktorklang replied:
Move out `reduceNext` as a private[this] method in the object, you mean?

viktorklang commented:
@retronym I addressed 8849 and added code-sharing between reduceLeft and foldLeft

retronym commented:
Can you please split the fix for `SI-8849` to a separate commit?

viktorklang committed:
SI-8849 Makes `ExecutionContext.Implicits.global` ambiguous

There was an unfortunate side-effect from having `Implicits.global` be
of type `ExecutionContextExecutor`; it is more specific than `ExecutionContext`,
as such it would be picked over other `ExecutionContexts` in the implicit scope.

retronym commented:
Could you also please take a look at:

https://issues.scala-lang.org/browse/SI-8689
https://issues.scala-lang.org/browse/SI-8615
https://issues.scala-lang.org/browse/SI-7846

To see if they are still relevant or if they offer ideas that could contribute to this PR?

retronym commented:
The fix for SI-8849 should be accompanied by a `neg` test to show that the use case in the ticket now issues the ambiguous implicit error.

Our commit message format is:

```
SI-1234 one line description of the change

Supporting text (importantly the whys and the consequences.)
```

viktorklang commented:
@retronym  see: https://github.com/viktorklang/scala/commit/3d19d0d0115f36099c93b49aac36675e0adab6a3

viktorklang commented:
https://issues.scala-lang.org/browse/SI-7846 <- This is "won't fix" as a result of this PR

https://issues.scala-lang.org/browse/SI-8615 <- can be easily implemented as

def retry(times: Int)(f: => Future[T])(implicit executor: ExecutionContext): Future[T] =
   f recoverWith { case t if times > 0 => retry(times - 1, f) }

I'm not sure that adding that method is justified in terms of usage, but I would be glad to be proven wrong.

https://issues.scala-lang.org/browse/SI-8689 <- I'll see if I can come up with a reproducer

retronym commented:
Thanks! Closing as won't fix is a perfectly fine resolution.

viktorklang commented:
Also added `zipWith` to Future (with tests & docs)

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -459,40 +514,106 @@ object Future {
   */
-  def apply[T](body: =>T)(implicit @deprecatedName('execctx) executor: ExecutionContext): Future[T] = impl.Future(body)
+  def apply[T](body: =>T)(implicit @deprecatedName('execctx) executor: ExecutionContext): Future[T] =
+    unit.map(_ => body)    
```
I'm not 100% sure about this change as it means that the creation of new futures are now `OnCompleteRunnable`s which means that they will get batched by batching executors. Thoughts @phaller?

phaller replied:
Excellent point, @viktorklang. I'll mull over it.

viktorklang replied:
Thanks @phaller!

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -354,17 +387,35 @@ trait Future[+T] extends Awaitable[T] {
+   *  @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
+   */
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
+    flatMap(r1 => that.map(r2 => f(r1, r2)))(internalExecutor)
```
Like I mentioned [on Twitter](https://twitter.com/philippkhaller/status/520558482282807296), in my opinion `zipWith` doesn't bring enough value to justify an increased API footprint. The following two are equivalent:

``` scala
thiz.zipWith(that)(f)
```

``` scala
thiz.zip(that).map(f.tupled)
```

Instead of adding to the API, maybe this could be an example in the docs?

softprops replied:
+1

retronym replied:
Is there an efficiency argument for this version? I'm thinking more in terms of number of suspend/resumes, than in the creation of the intermediate tuple.

viktorklang replied:
Dear @phaller & @retronym,

There -is- an efficiency argument compared to zip + map: it can do root-promise linking which is not possible with zip+map. We also avoid the extra allocation of the tupled version of the `f` and also the cost of the indirection of the tupled `f`.

Compared to user-site: `this.flatMap { r1 => other.map { r2 => f(r1, r2) } }` we gain some efficiency by using the `internalExecutor` since we control the logic that will be executed for the flatMap stage.

Although I both agree with the `don't bloat the API`-argument as well as `it's easy to implement yourself` the arguments above show that it is not, from a performance perspective, easy to do so.

Also, keep in mind that -most- methods on future are simple shims on top of `transform` and `transformWith`, so the `it's simple to implement yourself argument` would in that case argue to deprecate most methods on `Future` and "only" keep `transform` and `transformWith`—an argument that I don't think holds for the reasons mentioned above.

That being said, I am not -personally- invested in `zipWith`—I just thought it was missing.

phaller commented on src/library/scala/concurrent/impl/AbstractPromise.java:
```diff
@@ -34,7 +34,4 @@ protected final boolean updateState(Object oldState, Object newState) {
     }
-
-    protected final static AtomicReferenceFieldUpdater<AbstractPromise, Object> updater =
-	AtomicReferenceFieldUpdater.newUpdater(AbstractPromise.class, Object.class, "_ref");
```
Wow, surprised this was still there.

viktorklang replied:
+1000, I can't believe that it slipped through in the first place.

softprops commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -459,40 +513,100 @@ object Future {
 
+  /** A Future which is never completed.
+   */
+  final object never extends Future[Nothing] {
```
Just throwing this out there because I'm a fan of these kinds of constants, but have you looked at twitter util's [ConstFuture](https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Future.scala#L1101-L1141)? It's what they implement their  [constants](https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Future.scala#L19-L22) in terms of. `Nothing` probably covers all cases but I wanted to link to this here in case it was relevant. I love the choice of name for this btw :)

som-snytt replied:
I miss "Optimized for never doing anything" and I want the t-shirt, too.

paulp replied:
I'll give you a hint... [It involves the Internet](http://www.azlyrics.com/lyrics/barenakedladies/neverdoanything.html).

viktorklang replied:
@softprops Thanks for taking the time to comment! I suspect `ConstFuture` is ~[KeptPromise](https://github.com/scala/scala/blob/2.11.x/src/library/scala/concurrent/impl/Promise.scala#L322)

@som-snytt "Optimized for never doing anything" Is indeed an excellent t-shirt quote.

softprops commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -553,14 +729,38 @@ object Future {
+  }
+
   /** Transforms a `TraversableOnce[A]` into a `Future[TraversableOnce[B]]` using the provided function `A => Future[B]`.
    *  This is useful for performing a parallel map. For example, to apply a function to all items of a list
```
Wonder if it would be a good time to clarify this, but I recently had to do some head scratching here over whether or not this was really parallel vs async. Wasn't sure if ultimately it depended on the impl of the provided execution context.

viktorklang replied:
@softprops Good question. I've wanted to steer clear of "parallel" since to me it's about execution. ("Parallelizable" is closer to the semantics). Are you talking specifically about `transform` or in general in the `Future`-companion?

softprops replied:
In general I suppose. Maybe parallelizable is better. It may help avoid some confusion.

viktorklang replied:
@softprops I really like that idea. Let me have a stab at it!

viktorklang commented:
Update: I've also added `toString` representations to the stdlib Future implementations as well as added a performance improvement for already completed Futures (`Future.successful(…)` and `Future.failed(…)`) since the branches that will never be taken can be treated as no-ops.
I added tests to confirm those behaviors as well as verify the `toString` representation.

jozic commented on src/library/scala/concurrent/BlockContext.scala:
```diff
@@ -53,9 +53,16 @@ object BlockContext {
 
-  /** Obtain the current thread's current `BlockContext`. */
+  /**
+    @return the BlockContext that would be used for the current Thread at this point
```
a small nitpicking - this line misses \* at the beginning and BlockContext is not surrounded with back ticks as in other places, like line 75

viktorklang replied:
Thanks, fixed.

jozic commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -184,22 +189,24 @@ trait Future[+T] extends Awaitable[T] {
    *  Blocking on this future returns a value if the original future is completed with an exception
    *  and throws a corresponding exception if the original future fails.
```
could this be also rephrased?

viktorklang replied:
Reworded as:

```
/** The returned `Future` will be successfully completed with the `Throwable` of the original `Future` if the original `Future` fails.
   *  If the original `Future` is successful, the returned `Future` is failed with a `NoSuchElementException`.
   *
   * @return a failed projection of this `Future`.
   */
```

jozic commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -208,33 +215,49 @@ trait Future[+T] extends Awaitable[T] {
+  def transform[S](s: T => S, f: Throwable => Throwable)(implicit executor: ExecutionContext): Future[S] = 
+    transform {
+      case Success(r) => Try(s(r))
+      case Failure(t) => Try(throw f(t)) // will throw fatal errors!
```
Could we be more explicit about `// will throw fatal errors!` in scaladoc?

viktorklang replied:
It is not a concern to the scaladoc,
the reason for the throw is to make sure that fatal errors are rethrown to the transform, which can then handle the errors as always. So it is there to ensure that it behaves consistently with other uses of the method.

jozic replied:
i thought this case is a little bit different as usually all functions return a Try or a Future, but here user explicitly returns another Throwable. so here I meant we can probably explicitly mention what happens if a user converts a NonFatal throwable to a fatal one

jozic commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -208,33 +215,49 @@ trait Future[+T] extends Awaitable[T] {
+   *
+   *  @tparam S  the type of the returned `Future`
+   *  @param  f  function that transforms the result of this future
+   *  @return    a future that will be completed with the transformed value
```
should it be `a``Future``that will be ...` as it is on line 221 for example?

viktorklang replied:
yes, that was overlooked.

jozic commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -269,14 +294,15 @@ trait Future[+T] extends Awaitable[T] {
-    map {
-      r => if (p(r)) r else throw new NoSuchElementException("Future.filter predicate is not satisfied")
-    }
+    map { r => if (p(r)) r else throw new NoSuchElementException("Future.filter predicate is not satisfied") }
```
would be nice if a user can provide his/her own failure message, like

```
def filter(@deprecatedName('pred) p: T => Boolean, failureMessage: String = "Future.filter predicate is not satisfied")(implicit executor: ExecutionContext): Future[T] 
map { r => if (p(r)) r else throw new NoSuchElementException(failureMessage) }
```

but not sure about compatibility

viktorklang replied:
If the message of the exception is important, then one can always use `transform` to create a new exception with a new message.

jozic commented on src/library/scala/concurrent/Promise.scala:
```diff
@@ -76,7 +72,9 @@ trait Promise[T] {
    */
   final def tryCompleteWith(other: Future[T]): this.type = {
-    other onComplete { this tryComplete _ }
+    if (other ne this.future) { // this completeWith this doesn't make much sense
```
`// this tryCompleteWith this doesn't make much sense`?

viktorklang replied:
Sure!

jozic commented on src/library/scala/concurrent/impl/ExecutionContextImpl.scala:
```diff
@@ -96,56 +148,42 @@ private[scala] class ExecutionContextImpl private[impl] (es: Executor, reporter:
+      case anything: Throwable ⇒
+        val t = Thread.currentThread
+        t.getUncaughtExceptionHandler match {
+          case null ⇒
```
should `⇒` be replaced with `=>` for consistency?

viktorklang replied:
Absolutely, thank you!

jozic commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -85,6 +152,49 @@ class FutureTests extends MinimalScalaTest {
+        case _ => fail("Future.never did not match Future.`never`")
+      }
+
+      assert(test eq Future.never, "Future.unit must be the same instance as Future.unit")
```
`Future.never must be the same instance as Future.never`

viktorklang replied:
:+1:

jozic commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -218,6 +328,142 @@ class FutureTests extends MinimalScalaTest {
       } mustBe (r)
     }
 
+    "transform results to results" in {
```
are there tests for overloaded version with two functions?

viktorklang replied:
Yes, they are in the scala-concurrent-tck.scala file.

viktorklang commented:
Alright @retronym, I think this one is ready to go. I'd love for this to get into an early 2.12 milestone so we can collect Real World (tm) feedback and iron out any issues early.

viktorklang commented:
One thing that is left to do is to switch over to ForkJoinPool.commonPool() but that can only be done once we have switched over 2.12.x to Java8.

viktorklang commented:
Fixed a regression (t7146)

retronym commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 60015903)
:cat: Roger! Rebuilding pr-scala for fa0743c3, 44953dcb. :rotating_light:

retronym commented:
I would have preferred to defer the deprecation of the ForkJoin backport until we are building with Java 8 ourselves, so as to avoid deprecation warnings in our own codebase in the interim. But we'll do that pretty soon so I suppose we can let that pass.

LGTM

viktorklang commented:
@retronym Sounds good. So merge?

adriaanm commented:
Yep, :cake: for everybody!
